contrib/util.arrai

```diff
@@ -0,0 +1,112 @@
+# A collection of helper functions for arr.ai.
+#
+# If generally useful, these should gradually migrate to a more standard library.
+
+# Invokes a macro on a string as if it were source code at parsing time.
+let invokeMacro = \macro \s
+    macro -> (//dict(.@transform) >>> \rule \fn
+        fn(//grammar.parse(.@grammar, rule, s))).@value
+;
+
+# Transforms an AST into a simple tuple of its values.
+# Useful for the @transform of a flat grammar.
+let rec simpleTransform = \ast
+    cond ast {
+        (...):
+            let d = //dict(ast) >> \term cond term {
+                ('':value): value,
+                (...): simpleTransform(term),
+                [...]: term >> simpleTransform(.) where .@item,
+                _: {},
+            };
+            //tuple(d where .@value),
+        _: {}
+    }
+;
+
+# Filters the nodes of a hierarchical data structure based on a (key, value) predicate.
+# Key-value pairs for which the predicate returns false will be removed from the result.
+let rec filterTree = \pred \ast
+    cond ast {
+        {(@:..., @value:...), ...}: ast where pred(.@, .@value) >> filterTree(pred, .),
+        [...]: ast >> filterTree(pred, .),
+        {...}: ast => filterTree(pred, .),
+        (...): //tuple(//dict(ast) where pred(.@, .@value) >> filterTree(pred, .)),
+        _: ast,
+    }
+;
+
+# Sequentially applies `fn(accumulator, i)` for each `i` in `arr`. The `accumulator` is initialised
+# to `val`, and updated to the result of `fn` after each invocation.
+# Returns the final accumulated value.
+let rec reduce = \arr \fn \val
+    cond arr {
+        [head, ...]:
+            let tail = -1\(arr without (@:0, @item:head));
+            reduce(tail, fn, fn(val, head)),
+        _: val,
+    }
+;
+
+# Sequentially applies `fn(accumulator, k, v)` for each `{k: v}` pair in `obj`.
+# The `accumulator` is initialised to `val`, and updated to the result of `fn`
+# after each invocation.
+# Returns the final accumulated value.
+let rec reduceObj = \obj \fn \val
+    cond obj {
+        {}: val,
+        [(@:k, @value:v), ...tail]: reduceObj(tail rank (:.@), fn, fn(val, k, v)),
```

Comment:
does this actually work? shouldn't it be
```arrai
{(@: k, @value: v), ...tail}: ...
```

Reply:
It does, because a regular dict _won't_ match, and hence fall into `reduceObj(obj orderby .@, fn, val),`, which turns it into something of this shape upon recursing. It's kind of a hack to ensure a deterministic traversal order.